### ISSUE TITLE
[Feature Request] GCS and S3 support on windows

### ISSUE THREAD

**Issue by gunan** [tensorflow/tensorflow#19297]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Will be required.
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Windows, all versions
- **TensorFlow installed from (source or binary)**: source and binary,
- **TensorFlow version (use command below)**: master, and all releases
- **Python version**: python 3+
- **Bazel version (if compiling from source)**: - 
- **GCC/Compiler version (if compiling from source)**: - 
- **CUDA/cuDNN version**: - 
- **GPU model and memory**: - 
- **Exact command to reproduce**: Install tf on windows, and try anything that reads from GCS or S3

### Describe the problem
Currently, TF has no support for GCS and S3 on windows. Because they depend on curl, and we have not worked to make curl build and work on windows. Someone needs to dive in and work through the problems, and the rest should be just removing the windows exceptions for GCS and S3 support in configure.py.

This issue is open for community contributions.

**Comment by jart** [comment]
Here's an interesting fact I learned recently, which might help the person who implements this warn users about network egress fees, or misconfigured deployments. It's possible to automatically get the GCE host region with `curl -sfL metadata.google.internal/0.1/meta-data/zone | sed -n 's!.*/\([^-/]*-[^-/]*\).*!\1!p'` and get the GCS $BUCKET region prefix with `curl -sfLH "Authorization:Bearer $(curl -sfLH Metadata-Flavor:Google metadata.google.internal/computeMetadata/v1/instance/service-accounts/default/token | sed -n 's/.*"access_token": *"\([^"]*\)".*/\1/p')" "https://www.googleapis.com/storage/v1/b/$BUCKET" | sed -n 's/.*"location": *"\([^"]*\)".*/\1/p' | tr A-Z a-z`. If the DNS lookup fails, or the first string doesn't start with the second, then GCS egress isn't free. This is especially helpful for TensorBoard users, since it does bandwidth intensive log reading.

**Comment by vnvo2409** [comment]
I am interested in this feature and I am working on my proposal for this feature. Could you please provide more detail. Thank you 😄

**Comment by yongtang** [comment]
@vnvo2409 I can help you on this issue with respect to S3. If I remember, there were several dependencies that needs to be resolved for S3. curl is one, boringssl could be another one. In [tensorflow/io](https://github.com/tensorflow/io) we get the curl working in windows so you may copy the BUILD file and give it a try.

On a side note, there are ongoing discussions about moving S3/GCS support to [tensorflow/io](https://github.com/tensorflow/io) once modular file system C API is in place, @vnvo2409 maybe we should start the effort from tensorflow/io? Once the build is in place, we could looking into moving to modular file system. /cc @mihaimaruseac @gunan for input.

**Comment by mihaimaruseac** [comment]
Let's keep this on hold for a while longer until Windows modular filesystem is released (if all goes well, at most 3-4 weeks?)

**Comment by vnvo2409** [comment]
@yongtang I wrote a small C++ program to list the content of a bucket on S3 ( by sending a curl http get request) and it works with openssl and the libcurl provided by Windows. However, I did not manage to build the curl by [tensorflow/io](https://github.com/tensorflow/io) on window. Could you please provide more instructions about building it on windows.

@mihaimaruseac If the modular filesystem is released and goes well, will we continue working on this feature ( by resolving the missing dependencies ) or we will write plugins to provide S3 and GCS support. If there is any change, please let me know because the deadline to submit proposal for GSoC will be 31th March. 

And finally, beside sending my draft to webpaige@google.com, is there anyone could provide feedbacks for me. Thank you 😄

**Comment by yongtang** [comment]
@vnvo2409 Building [tensorflow/io](https://github.com/tensorflow/io) on windows is done through Bazel, we have GitHub Actions CI to check every commit: https://github.com/tensorflow/io/actions

The config file of GitHub actions is located: https://github.com/tensorflow/io/blob/master/.github/workflows/build.yml

Inside you can find related Windows part:
https://github.com/tensorflow/io/blob/edd02b2cefd1dbf719990b42ea43de4c4eede2b5/.github/workflows/build.yml#L241-L270

Setting up Bazel on Windows is not exactly straightforward, (that is also why the the issue of porting GCS/S3 to Windows is still open). But we are here to resolve it.

@vnvo2409 Could you open an issue in [tensorflow/io](https://github.com/tensorflow/io)  if you encountered any issues in building it on Windows?

**Comment by mihaimaruseac** [comment]
@vnvo2409 once we have modular filesystem support for windows, we will start replacing the current ones with plugins for GCS, S3, Hadoop. Plan is for these plugins to reside under [tensorflow/io](https://github.com/tensorflow/io) and to work from the start on Linux, Mac and Windows.

Note that there are also two other students that are supposed to start working on filesystems as part of [ROSEdu](http://www.rosedu.org/en/) CDL (open source development course)

**Comment by vnvo2409** [comment]
I built tensorflow on windows with s3 successfully by removing the line `clean_dep("//tensorflow:windows"): [],` in `tensorflow/core/platform/default/build_config.bzl`. However, when I tried to test, the exceptions 'file scheme is not implemented' are remains. Could someone please tell me how to remove the exceptions ? Thank you :smile:

**Comment by mihaimaruseac** [comment]
Interesting. It shouldn't use the `tensorflow/core/platform/default` paths, but the `tensorflow/core/platform/windows` ones if on Windows.

**Comment by vnvo2409** [comment]
I ask @yongtang  and he told me to remove this line https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/platform/default/build_config.bzl#L656-L660 and It trigger the s3 builds on windows. Anyway, I think that It is built succesfully. However, the exceptions "File system S3 is not implemented" are remains. Do you @mihaimaruseac know where to remove the exceptions ?
In addition, I am considering to change my proposal idea for Google Summer Of Code into write the plugins S3/GCS for your modular filesystem, could you please tell me which filesystems those 2 students work on ?
Thank you 😄

**Comment by yongtang** [comment]
@vnvo2409 The `"File system S3 is not implemented" ` error might be caused by the fact that `s3://` schema is not registered. Beside building successfully the schema also need to be registered. I haven't touched this part of the code for quite sometime so thing might have changed. On a side note I think write plugins for modular file system C API would be better.

**Comment by vnvo2409** [comment]
When exploring the source code, I think that writing plugins for modular file system would be easier if we already had S3/GCS supports on Windows ( The plugins and the support are somehow similar, aren't they ?) So I will try to register `s3://` schema, could you tell me more detail how to register. Even if it is not so useful, it will be a great opportunity for me to learn more about tensorflow source code and bazel :smile:

**Comment by vnvo2409** [comment]
In addition, Could anyone please tell me who will be the mentor for the GCS and S3 supports on Windows ? It would be great if I could contact the mentor and have some feedbacks for my proposal. Thank you

**Comment by yongtang** [comment]
@vnvo2409 The registration code is in:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/platform/s3/s3_file_system.cc#L1039

So it should have been registered if the build is successful. You may need to trace more to see where the issue comes from.

**Comment by yongtang** [comment]
> In addition, Could anyone please tell me who will be the mentor for the GCS and S3 supports on Windows ? It would be great if I could contact the mentor and have some feedbacks for my proposal.

I don't know who was the original mentor that proposed S3/GCS on Windows. You can contact organizer and send email to tensorflow-developers (google group) email list. If no mentor, I could
sync up with organizer to see if I can be mentor. (Served as GSoC mentor last year but that was a different project).

**Comment by vnvo2409** [comment]
I think the issue may come from what @gunan say at the open of issues. I did dive into the configure.py but I found nothing
> and the rest should be just removing the windows exceptions for GCS and S3 support in configure.py.

**Comment by mihaimaruseac** [comment]
Sorry for late update, I went away for a while.

I will probably be the mentor for this project since I proposed the refactoring in tensorflow/community#101

The only lines with `clean_dep("//tensorflow:windows") : []` are [these 3](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/core/platform/default/build_config.bzl;l=638,647,656;drc=5473ad460d7f1831fac4b3882a37422bbc5c2aa9) and indeed, removing each one of them makes the specific filesystem be compiled on Windows. So, I was wrong above when I said that `platform/default/build_config.bzl` doesn't need to be changed. Apologies.

Filesystem not registered is because the file that contains this line has not been compiled on Windows (https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/platform/s3/s3_file_system.cc#L1039 as suggested by @yongtang ) Can you try removing [this windows deps](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/core/platform/default/build_config.bzl;l=656;drc=5473ad460d7f1831fac4b3882a37422bbc5c2aa9) and compiling again? You can also add a print [at this line](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/core/platform/env.cc;l=73;drc=8f597046dc30c14b5413813d02c0e0aed399c177) and print the `scheme` argument. That should tell you which filesystems are registered and in what order? Or, in a C++ code linking to TensorFlow, you can call [`Env::GetRegisteredFileSystemSchemes`](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/core/platform/env.cc;l=129;drc=8f597046dc30c14b5413813d02c0e0aed399c177). See an example [here](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/c/experimental/filesystem/modular_filesystem_test.cc;l=1685-1687;drc=23ff63d546800cd7143172bb201d6bb8d6fd15fd)

Regarding the modular API, we are still in flux now working on the Windows one. Every modular implementation should be at [tensorflow/c/experimental/filesystem/](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/c/experimental/filesystem/) Please look through the interface defined there and the implementation of the POSIX filesystem.

My other two students haven't yet started implementation of the modular filesystems, so we can start here instead, since I adhere to the philosophy of "whoever writes the code first gets the PR merged"

**Comment by vnvo2409** [comment]
@mihaimaruseac So I will start implementing the plugins right away. Could I send you my proposal and get some feedbacks about it please ?
Thank you :smile:

**Comment by vnvo2409** [comment]
I try `std::cout` the `scheme` argument and here is what I found
![Screenshot 2020-03-22 at 09 47 53](https://user-images.githubusercontent.com/17511625/77241298-9bfda280-6c22-11ea-95dd-4e824d6451f0.png)
It seems that the `s3://` file schema has already been implemented and the exceptions are hard-coded into some file ? I have to put the `#` to make it into a comment to prevent conflict with auto-generated file.
I will try the `Env::GetRegisteredFileSystemSchemes` and come back later

**Comment by mihaimaruseac** [comment]
I will look over the proposal.

The output seems to come from compiling the code. Question is what is the output when you run the program

**Comment by vnvo2409** [comment]
the output when I import tensorflow is the same.
![Screenshot 2020-03-22 at 10 47 18](https://user-images.githubusercontent.com/17511625/77241844-9441fc00-6c2a-11ea-8fed-ce9ee21fdf34.png)

However, the compile output of `Env::GetRegisteredFileSystemSchemes` does not give me `s3`. I will dive into it.

**Comment by mihaimaruseac** [comment]
That's interesting. Can you try again by doing `python -vvv -c "import tensorflow" &> log` and then sharing the entire `log` file?

**Comment by vnvo2409** [comment]
I redirection all the outputs and errors so it is a huge file ( 6mb ~ 68k line )

[python_log.txt](https://github.com/tensorflow/tensorflow/files/4364664/python_log.txt)

**Comment by vnvo2409** [comment]
I think it might be useful. I put a static variable in the function `FileSystemRegistryImpl::Register` to count how many time it is called. It turn out that it is called by more than one process (?)
<img width="366" alt="image" src="https://user-images.githubusercontent.com/17511625/77252723-64294600-6c88-11ea-9752-27ac16fb6bf3.png">
the first process call it 3 times with "s3", "" and "file" ( the last two are for windows ). However, from the second process ( when the static variable is 1 ), "s3" is not registered anymore and therefore the exceptions happen.

**Comment by mihaimaruseac** [comment]
Yes, I was thinking to tell you to print the address of `registry_`, both in `FileSystemRegistryImpl::Register` and `FileSystemRegistryImpl::GetRegisteredFileSystemSchemes`

**Comment by mihaimaruseac** [comment]
Another thing to try: locate the dll of the s3 filesystem and then try to import it from python, via [`tf.load_library`](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/python/framework/load_library.py;l=123;drc=a02fe6c24a5a38c24d0330120c8426aeea28db9f)

**Comment by vnvo2409** [comment]
ok I will try. Thank you

**Comment by vnvo2409** [comment]
@mihaimaruseac There is no dll of s3 filesystem, there is only a s3_file_system.lo.lib which can not be loaded into tensorflow. I think that the problems might be come from `genrule` (you can see the call for `Register` is `from executing genrule`, as it needs to generate file, it call `FileSystemRegistryImpl::Register` for Windows and we lost the `s3://`. I think I will try to hard-code the `REGISTER_FILE_SYSTEM("s3", RetryingS3FileSystem);` into this file https://github.com/tensorflow/tensorflow/blob/f7b6793c6611210405d066dde84cc67adca4097c/tensorflow/core/platform/windows/env.cc#L193
<img width="810" alt="image" src="https://user-images.githubusercontent.com/17511625/77258524-b3ce3880-6cad-11ea-9a9d-efd702f10141.png">

**Comment by mihaimaruseac** [comment]
Yeah, every genrule is a different process which does another filesystem registration. So this makes sense so far.

It is interesting that there is no dll. I think first step would be to make the necessary build changes to create one. Probably we need a `tf_cc_shared_object` rule, like [I did for the modular windows plugin](tensorflow/c/experimental/filesystem/plugins/windows/BUILD)

**Comment by vnvo2409** [comment]
I can not figure out how to deal with the genrule. I can not build the dll too. However, there is a good news that @gunan said that he has compiled successfully gcs support on Windows https://groups.google.com/a/tensorflow.org/forum/#!topic/developers/wYlMfe3KIYk . I think the best thing I can do now is waiting for his reply about the exceptions. I will also change my proposal to the second approach as the first approach is nearly done. Could you review my proposal for me when it is done @mihaimaruseac ?
Thank you 😄

**Comment by mihaimaruseac** [comment]
Sounds good to me

**Comment by gunan** [comment]
Certainly a lot happened on this thread. I will try to summarize my responses.
In TFs codebase, now curl builds on windows. I have resolved all issues a couple weeks ago.
I finally also have tensorflow/core/platform/cloud:gcs_file_system_test passing, too.

about the changes to platform/default/build_config.bzl. You cannot simply remove the `tensorflow:windows": []` I think more than filesystems depend on that. You will need to look deeper into what build targets use these macros.

About the exceptions, as @yongtang mentioned, `File system S3 is not implemented` means S3 module is just not built. When the module is built into the binary, it registers the file name schema for s3, and the error goes away.

**Comment by gunan** [comment]
I think there may be an ODR violation, file system registry may be linked twice. Because even when I explicitly built GCS support, I seem to be missing it.

**Comment by vnvo2409** [comment]
So you have the same problems with me ? @gunan

**Comment by vnvo2409** [comment]
@mihaimaruseac I finished my proposal. Thank you for review it one more times :smile:

**Comment by vnvo2409** [comment]
I can not build `//tensorflow/core/platform/s3:s3_file_system_test` becasue of this error
<img width="1680" alt="image" src="https://user-images.githubusercontent.com/17511625/77523288-1dba2e00-6eb8-11ea-8193-ebea8d64ff72.png">
I think the root of it is there are two file [1](https://github.com/tensorflow/tensorflow/blob/0298dd6900d4278e1a3f8c4f6a6956dd9e7ca7f0/tensorflow/core/lib/bfloat16/BUILD#L13) and [2](https://github.com/tensorflow/tensorflow/blob/0298dd6900d4278e1a3f8c4f6a6956dd9e7ca7f0/tensorflow/core/framework/BUILD#L584) which both contain the rule to build `bfloat16.dll`
Could anyone help me ? Thank you

**Comment by mihaimaruseac** [comment]
Interesting. But, afaik, the two rules should generate two different files in two different paths.

I need to set-up a Windows environment and check

**Comment by gunan** [comment]
That is a known issue.
Due to path length restrictions, bazel has these name clashes on windows
when running bazel test.
You need to add "--dynamic_mode=off" flag to your bazel test commands on
windows to be able to tests

**Comment by vnvo2409** [comment]
I run the test and not a single test pass. However, there is a good news that the registry_ only linked one times and the `s3://` works. I think we have to edit the path to make it works on windows
<img width="1089" alt="image" src="https://user-images.githubusercontent.com/17511625/77599885-d9bb3d80-6f38-11ea-8b4c-58961202626d.png">

**Comment by vnvo2409** [comment]
Another question, could anyone show me how to set up a build trigger process ? I see that `tensorflow` using cloud build API by GCP but I am not sure how to set up for my own

**Comment by vnvo2409** [comment]
Good news, S3 has passed all unit test except `HasAtomicMove` which does not compile because `NeedTempsLocation` is not belong to `s3fs`. I have to set up my own bucket on S3 sever to test. Is there a way to mock test it ?
Anyway, the only thing left to do is solving the exceptions 🚀 
<img width="809" alt="image" src="https://user-images.githubusercontent.com/17511625/77730733-f6886b80-7033-11ea-94a0-f9331d6def21.png">

**Comment by mihaimaruseac** [comment]
That's great.

We would prefer to not mock the filesystem tests because that could result in divergent behavior between actual use and the mock.

Regarding the missing method, that might be because when it was added recently no one also ran the S3 tests.

**Comment by vnvo2409** [comment]
Ok. I am working on the exception and I will update when I find something ?
Do you have any suggetion where to dig in for me ? @mihaimaruseac

**Comment by yongtang** [comment]
@vnvo2409 In tensorflow-io we use [localstack](https://github.com/localstack/localstack), which is a AWS stack emulator, to test AWS related ops. It has been quite useful to us. This might be suitable for tensorflow as tensorflow's code base and test running time is already quite long. But once s3 is migrated to tensorflow-io we could reuse localstack to setup the test for s3.

**Comment by vnvo2409** [comment]
![image](https://user-images.githubusercontent.com/17511625/77824815-ec4f9580-7137-11ea-80b9-0268f34c7227.png)
I see that the `registry_` used by `Env::GetFileSystemForFile` is the fifth, not the very last one as I thought ( I've already tried many times )
@mihaimaruseac I think that I need help now because it seems very werid to me as I think that the next `registry_` should override the last one. I could not get any further debug information. Could you explain the order of linking `registry_` for me 
Thank you.

**Comment by mihaimaruseac** [comment]
In theory it should be just one global variable. But because it is used across multiple shared objects, there is a possibility that one shared object gets a `registry_` initialized at a different address. The keyword here is ODR violation (one definition rule). We need to debug the build and find out which DLLs provide the alternate definition

**Comment by vnvo2409** [comment]
Good news every one. I have made `s3` worked on Windows. I have to add to this rule `deps`
```
select({
        "//tensorflow:windows": ["//tensorflow/core/platform/s3:s3_file_system"],
        "//conditions:default": [],
})
```
https://github.com/tensorflow/tensorflow/blob/907a55ebadce3c78bef73148613b327ed84d92e4/tensorflow/python/BUILD#L6070-L6081
<img width="560" alt="image" src="https://user-images.githubusercontent.com/17511625/78277615-00800200-753f-11ea-9b19-0e740a73ae40.png">

@mihaimaruseac Now I will make one PR for `s3` and another ones for `gcs` or I should make two filesystems in one PR ?

**Comment by mihaimaruseac** [comment]
This addition works, but it has a downside. Now the filesystem is linked always on Windows, even for users who might not want to use S3.

Preferably we should provide a `tf_cc_shared_object` rule to build the DLL of the S3 filesystem on Windows (and an `.so` on POSIX) and then in Python we should use `load_library` to load the plugin on demand.

**Comment by vnvo2409** [comment]
I build tensorflow on Linux and `s3` is also loaded when we `import tensorflow`. So I think this addition is more suitatble for the tensorflow source code right now. About the idea of building the DLL of the S3 filesystem, it is more suitable when we finish the implementation of filesystem modular. Because right now, in the source code of tensorflow, we always register a filesystem by a macro even for users who do not use s3.
https://github.com/tensorflow/tensorflow/blob/ca6e96effcd3d3bcb9eec96e09832519b1acda70/tensorflow/core/platform/s3/s3_file_system.cc#L1039
I think also find the cause of ODR violation, it is because on windows `tensorflow` use 2 folder `core/platform/default` and `core/platform/windows` instead of only `core/platform/default` like on Linux. To solve the ODR violation, I think we have to re-organize the 2 folder.
@mihaimaruseac  I would like to ask you, what should I do, try to solve the ODR violation on Windows or begining to work on the plugins for modular filesystem ?

**Comment by vnvo2409** [comment]
I think I finally found the root cause of ODR violation. The problem is about the different of bazel config between Windows and Linux. This line is default on Windows but it is not on Linux
https://github.com/tensorflow/tensorflow/blob/5daf0e00e9913a0f7bd262f5f37ea70d34b00811/.bazelrc#L119

On linux, we have a shared object named `libtensorflow_framework.so.` which hold a common instance of `Env::Default()` and is imported into all shared objects that require a static instance of `Env::Default()`. So If there is a need of accessing to filesystem, it can use the common `Env::Default()` rather then creating a new instance.

On Windows, that line is considered into the bazel build config, which means it does not build `libtensorflow_framework.dll`. The `_pywrap_tensorflow_internal.so` is imported into these shared objects, but it does not export the symbols `Env::Default()` so these shared objects must create a new instance of `Env::Default()`.

**Comment by vnvo2409** [comment]
On Linux, when I build with `--config=monolithic`, the behavior is extractly the same on Windows. `s3` is not linked into tensorflow and there are several ODR violations, just like windows

**Comment by mihaimaruseac** [comment]
I think the plugin has to be loaded manually via `load_library`. At least on Linux the shared object can be built.

**Comment by vnvo2409** [comment]
@mihaimaruseac Could you tell me more about the current implementation of C API Modular Filesystem. So that I can have a plan working on the plugin and not to put too much efforts here

**Comment by mihaimaruseac** [comment]
That sounds good.

The entire implementation is under [`c/experimental/filesystem`](https://osscs.corp.google.com/tensorflow/tensorflow/+/master:tensorflow/c/experimental/filesystem/). There is a [test suite](https://osscs.corp.google.com/tensorflow/tensorflow/+/master:tensorflow/c/experimental/filesystem/modular_filesystem_test.cc), the [API interface plugins needs to be compiled against](https://osscs.corp.google.com/tensorflow/tensorflow/+/master:tensorflow/c/experimental/filesystem/filesystem_interface.h) (with a lot of comments) and some other glue files.

The plugins are under [`c/experimental/filesystem/plugins`](tensorflow/c/experimental/filesystem/plugins/). POSIX is fully implemented as of now, Windows is just the starting stub, but expect more changes to come there in the coming weeks.

Please read at the least the [API](https://osscs.corp.google.com/tensorflow/tensorflow/+/master:tensorflow/c/experimental/filesystem/filesystem_interface.h) (takes precedence over tensorflow/community#101) and let me know if there are questions

**Comment by vnvo2409** [comment]
Everything you send me require me a MOMA Single Sign On and I do not have the account 😢

**Comment by yongtang** [comment]
@vnvo2409 The files are located on the master branch of the tensorflow repo. You can replace name mentioned by @mihaimaruseac with the repo, e.g., :

https://github.com/tensorflow/tensorflow/tree/master/tensorflow/c/experimental/filesystem

**Comment by vnvo2409** [comment]
thank you

**Comment by mihaimaruseac** [comment]
Oh, sorry, I was using the links from my work computer. You can also replace the prefix of the links from `osscs.corp.google.com`, to `cs.opensource.google`

**Comment by vnvo2409** [comment]
@mihaimaruseac Can I use `AWS sdk for C++` for `s3` plugin or I should use as much `C` as possible ? A small note, I have to delete 2 lines to able to test `libposix_filesystem.so` 
https://github.com/tensorflow/tensorflow/blob/3364a141f2a7dc151154652e2e231516d6ee43cc/tensorflow/core/platform/default/env.cc#L215-L216

**Comment by mihaimaruseac** [comment]
Apologies for the delay in answering.

Regarding the deletion of the two lines: yes, that is expected. Otherwise the filesystem in use is the old one. See #37734 for the latest updates w.r.t. testing. I should actually create a guide for how to test these but was planning on doing that after Windows support was ready.

Regarding language usage: the plugin could use C++ and libraries but the interface must be valid C.

**Comment by vnvo2409** [comment]
@mihaimaruseac I have one more question, Could I reuse some part of tensorflow `core` or not ? I see the current implementation of s3 filesystem is heavily depended on the lib in `tensorflow/core/`. I think it is not fit to the principle of Modular Tensorflow.

**Comment by mihaimaruseac** [comment]
Regarding 1: preferably not because then the plugin and the core of TensorFlow will still be entwined to the same compiler version and compile flags, breaking any guarantee that we want to achieve with the modularization. Instead, we should identify what are the common things and expose those to an interface, _if_ there are multiple users.

Regarding 2: We would prefer to not initialize static global variables which are not trivially destructible. Do we need `shared_ptr` on the types? Keep in mind that the opaque structs in the API can be used to hide plugin specific implementation on the plugin side

**Comment by vnvo2409** [comment]
I think we need something to keep a single instance of variable of type `S3Client` to use when s3 plugins is loaeded because allocating a new `S3Client` each time we need would be a waste of resource.

**Comment by mihaimaruseac** [comment]
Agree. But the plugin is only loaded a single time, it cannot be unloaded

**Comment by vnvo2409** [comment]
So I suggest using a `std::weak_ptr`
Something like
```cpp
// s3_filesystem.cc
namespace tf_random_access_file {

typedef struct S3File {
	const char* bucket_;
	const char* object_;
	std::shared_ptr<Aws::S3::S3Client> s3_client_;
};
}

// s3_helper.h
namespace tf_s3_filesystem {
  std::shared_ptr<Aws::S3::S3Client> GetS3Client();
  std::shared_ptr<Aws::Transfer::TransferManager> GetTransferManager();
  std::shared_ptr<Aws::Utils::Threading::PooledThreadExecutor> GetExecutor();
}  // namespace tf_s3_filesystem

// s3_helper.cc  
// use anonymous namespace to hide these pointers
namespace tf_s3_filesystem {
  namespace {
    std::weak_ptr<Aws::S3::S3Client> s3_client_;
    std::weak_ptr<Aws::Transfer::TransferManager> transfer_manager_;
    std::weak_ptr<Aws::Utils::Threading::PooledThreadExecutor> thread_executor_;
  }

  std::shared_ptr<Aws::S3::S3Client> GetS3Client() {
    if(s3_client_.get() == nullptr ) {
      //...
      s3_client_ = std::shared_ptr<Aws::S3::S3Client>(...);
    }
    return s3_client_;
  }
}  // namespace tf_s3_filesystem
```
The global static will be reference by a `weak_ptr<Aws::S3::S3Client>`. If something need to use it, we could return a `shared_ptr<Aws::S3::S3Client>`. When it is no longer use anymore, It will be destructible right at that moment, and we don't have to keep it throughout the program as the plugin can not be unloaded.

**Comment by mihaimaruseac** [comment]
This looks like something that should be done in the function passed to [`init`](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/c/experimental/filesystem/filesystem_interface.h;l=286-289;drc=12d516b9cd7a7954fa79cccfe645564837cc4d23).

In the plugin, you can have a private struct that keeps these pointers (naked pointer should be ok, I don't see a reason to have them be `std::weak_ptr` or other wrappers at the moment) and then you [attach that struct to the opaque `TF_Filesystem`](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/c/experimental/filesystem/filesystem_interface.h;l=77-79;drc=12d516b9cd7a7954fa79cccfe645564837cc4d23).

For example, [POSIX `init`](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/c/experimental/filesystem/plugins/posix/posix_filesystem.cc;l=191-193;drc=ac271534b897ea2bb78a865971823f814b0151e1) does nothing, but you could have something like

```cpp
namespace tf_s3_filesystem {

typedef struct S3Data {
    Aws::S3::S3Client *s3_client;
} S3Data;

static void Init(TF_Filesystem* filesystem, TF_Status* status) {
  auto s3data = static_cast<S3Data*>(filesystem->plugin_filesystem);
  s3data->s3_client = ...;
  TF_SetStatus(status, TF_OK, "");
}

}  // namespace tf_s3_filesystem
```

Then, when you create a file:

```cpp

namespace tf_s3_random_access_file {

typedef struct S3File {
	const char* bucket;
	const char* object;
	const Aws::S3::S3Client *s3_client; // not owned
};
}  // namespace tf_s3_random_access_file

namespace tf_s3_filesystem {

static void NewWritableFile(const TF_Filesystem* filesystem, const char* path,
                            TF_WritableFile* file, TF_Status* status) {
  auto s3data = static_cast<S3Data*>(filesystem->plugin_filesystem);
  ...
  file->plugin_file = new tf_random_access_file::S3File({bucket, object, s3data->s3_client})
  TF_SetStatus(status, TF_OK, "");
}

}  // namespace tf_s3_filesystem
```

This way you don't need globals.

**Comment by vnvo2409** [comment]
Thank you. I have misunderstood the usage of `const TF_Filesystem* filesystem`

**Comment by mihaimaruseac** [comment]
Regarding the pointers, yes, if the API already uses wrapped ones then we should incorporate them too. I was not checking the AWS API.

Regarding the second point: that's great, I like the lazy loading approach you suggested.

**Comment by vnvo2409** [comment]
Here is what I have done with the s3 plugins

1. I have created a `s3` filesystem plugins based on the current implementation of `s3` filesystem.
Here is the test log. [attempt_1.log](https://github.com/tensorflow/tensorflow/files/4503596/attempt_1.log). As you can see. **94/103** test passed. [Here](http://bucketvnvo.s3.amazonaws.com) is the link to my test bucket on s3. I have removed some test: 
- `TestTranslateName` because the default `TranslateName` remove the bucket name ( `s3://bucket/tmp/test` -> `/tmp/test/`) So I have to write my own `TranslateName`.
- Any test related to `MemoryRegion`. `Seg fault` so I will work on it later.
- `TestStatPathIsInvalid` I use `Stat` to check if path is invaild or not so this test can not be run.

2. Test time is very long. It is because most of the tests are not compability with how `s3` filesystem work and they just add the complexity to the operation of `s3` plugins filesystem. `s3` does not have the `directory` concept. `s3` has a flat structure instead of a hierarchy like in a normal filesystem. Take an example, with only one bucket `bucketvnvo` without any "directory"
```bash
aws s3api put-object --bucket bucketvnvo --key test/vnvo/file
aws s3api put-object --bucket bucketvnvo --key test/vnvo/file/dir
```
These 2 command create 2 objects with have the key `test/vnvo/file` and `test/vnvo/file/dir`. It does not care if the "parent directory" is exist or point to a file or something as there is no "parent directory". So `TestCreateFileNonExisting`, `TestCreateFileExistingDir`, `TestCreateFilePathIsInvalid`, etc are not so useful. I will remove some `path pre-checker` in the `s3` plugins filesystem and I think we should rewrite the modular filesystem test for `s3` 

A side note, I have some exams so I will be less active till the student announcement (May 5) of the Google Summer of Code.

**Comment by mihaimaruseac** [comment]
Since S3 does not support directories those operations should return the not implemented status. Then tests would be marked as passing and skipped fast.

The goal for the tests is to provide a checker that all plugins adhere to the specification of the API in the modular interface.

Do you have a PR with the code so we can look through that?

No worries about the delay, good luck on the exams

**Comment by vnvo2409** [comment]
@yongtang I see you have fixed the `gcs_filesystem` linking problems on Windows by this PR https://github.com/tensorflow/tensorflow/pull/38959. Do you think I should make a PR for `s3_filesystem` on Windows as well ( it has already passed all tests ) and we can close this issue.

**Comment by mihaimaruseac** [comment]
That sounds good, @vnvo2409. Please assign/mention me in the PR.

**Comment by vnvo2409** [comment]
@mihaimaruseac I have made a PR, please review it.
https://github.com/tensorflow/tensorflow/pull/39124

**Comment by mihaimaruseac** [comment]
Closed by #39124
